DTLS socket and timeout fixes

julek-wolfssl opened the pull request:
- Handle `recvfrom()` 0 return. This indicates a closed socket.
- TCP sockets set `errno == SOCKET_EINTR` on a timeout
- Don't send alerts on a decryption error in DTLS
- We should always send alerts when TLS fails decryption

julek-wolfssl committed:
Reset timeout when reading a valid DTLS message

- Increment the DTLS 1.3 timeout on a long timeout

julek-wolfssl committed:
Silently discard DTLS msgs that fail decryption

Don't send alerts when decryption fails inside a DTLS connection.
TLS should always send a bad_record_mac when decryption fails.

julek-wolfssl committed:
EmbedReceiveFrom: fix when using a TCP socket

- recvfrom() returns 0 on a closed TCP socket
- TCP sockets set WOLFSSL_CBIO_ERR_ISR on a timeout

rizlik commented on src/wolfio.c:
```diff
@@ -473,11 +473,28 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+        int type;
+        XSOCKLENT length = sizeof(int); /* optvalue 'type' is of size int */
+
+        if (getsockopt(sd, SOL_SOCKET, SO_TYPE, &type, &length) == 0 &&
```
Now that we use this code twice it may be worthwhile to do a small static function to get the socket type.

julek-wolfssl replied:
Refactored in one func.

rizlik commented on src/dtls13.c:
```diff
@@ -2296,6 +2296,10 @@ int Dtls13RtxTimeout(WOLFSSL* ssl)
 
+    /* Increase timeout on long timeout */
+    if (DtlsMsgPoolTimeout(ssl) != 0)
+        return -1;
```
we error out once reached max timeout, so after about `max_timeout/DTLS_TIMEOUT_MULTIPLERS` numbers of retransmission. 

I think a better solution would be to explicitly set a maximum number of retransmissions than relying on the maximum timeout. I would just let the timeout saturate: stop its increase once it reaches the maximum. This means discarding the return value of `DtlsMsgPoolTimeout`. 

If you think that adding a explicitly counter of max retransmission is outside the scope for this PR leave it as it is for now so we have at least a limit that would stop hanging tests.

julek-wolfssl replied:
I think that this change should be introduced separately, but I also don't see much advantage in changing this to a repeat counter instead of a timeout limit.

>I would just let the timeout saturate: stop its increase once it reaches the maximum.

We need a way to error out on a non-responsive peer. How would you suggest we error out? Or do you mean that we introduce a separate re-transmission counter from the timeout counter?

rizlik replied:
Yes, I mean a separate re-transmission counter. After reaching MAX_NUMBER_OF_RTX we error out.

julek-wolfssl replied:
That would mean that we should re-think our timeout setting API as currently we set the initial value and the maximum value. Let's tackle this outside of this PR.

rizlik commented on src/wolfio.c:
```diff
@@ -473,11 +473,28 @@ int EmbedReceiveFrom(WOLFSSL *ssl, char *buf, int sz, void *ctx)
+
+        err = TranslateIoError(recvd);
+        if ((err == WOLFSSL_CBIO_ERR_WANT_READ ||
+                err == WOLFSSL_CBIO_ERR_ISR) &&
```
I'm not sure that TCP sockets set `WOLFSSL_CBIO_ERR_ISR` so the commit's changelog is slightly misleading. Nevertheless, I think it's correct to handle `WOLFSSL_CBIO_ERR_ISR` as the `WOLFSSL_CBIO_ERR_WANT_READ`. The only side-effect I see is that we may increase the timeout faster than we should in case of a signal interrupting the recvfrom.

julek-wolfssl replied:
We get a `WOLFSSL_CBIO_ERR_ISR` when we reach the timeout on a STREAM socket. I'm not sure why this is instead of a `WOLFSSL_CBIO_ERR_WANT_READ` but that is what is happening here.

rizlik replied:
I tried to test this with a toy c program but I get an `EAGAIN` (that should be translated to `WOLFSSL_CBIO_ERR_WANT_READ`) instead. Can you investigate further?

julek-wolfssl replied:
I could not replicate this behaviour. I'm not sure what I was seeing when I wrote this. I'm guessing the debugger might have been messing with the system call. Reverting back to how it was.

rizlik reviewed (changes requested):
Good work,
If the requested changes regarding the alerts on decryption failures get in the way it may be helpful to split this PR into two separate ones.

kaleb-himes commented:
retest this please

julek-wolfssl commented on src/internal.c:
```diff
@@ -18554,28 +18559,20 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
-                        #endif /* WOLFSSL_DTLS_DROP_STATS */
-                    }
-                #endif /* WOLFSSL_DTLS */
+                    SendAlert(ssl, alert_fatal, bad_record_mac);
```
Multiple alerts shouldn't be dangerous but we should try to stick to one fatal alert.

julek-wolfssl replied:
I refactored the alerts in the last commit. Now it gets sent in one place in `ProcessReply` instead of the `DecryptTls*` functions.

dgarske commented:
retest this please

julek-wolfssl committed:
Refactor checking socket type into a function

julek-wolfssl committed:
Refactor sending alert on decryption failure

Take sending of the alert outside of DecryptTls() and DecryptTls13(). The alert is now sent in ProcessReplyEx().

rizlik commented on src/tls13.c:
```diff
@@ -2278,11 +2278,10 @@ static int Tls13IntegrityOnly_Decrypt(WOLFSSL* ssl, byte* output,
  * sz      The length of the encrypted data plus authentication tag.
  * aad     The additional authentication data.
  * aadSz   The size of the addition authentication data.
- * doAlert Generate alert on error (set to 0 for sniffer use cases)
```
can this be a problem for the sniffer now?

julek-wolfssl replied:
Sniffer was also modified to use the new api.

julek-wolfssl commented:
Assigning to @SparkiDev as this touches some TLS 1.3 code.
